pkg/phlaredb/profile_store.go

```diff
@@ -0,0 +1,452 @@
+
+	// Upgrade profile getters to point to correct location
+	for pos, pg := range s.getters {
+		// TODO: Reuse the profile struct eventually by using a pool
+		pg.p = nil
+
+		pg.rowNum = int64(pos) + 1
+		pg.rowGroup = rowGroup
+	}
+
+	s.slice = s.slice[:0]
+	s.getters = s.getters[:0]
+	s.size.Store(0)
+
+	return nil
+}
+
+func (s *profileStore) writeRowGroups(path string, rowGroups []parquet.RowGroup) (n uint64, numRowGroups uint64, err error) {
+	fileCloser, err := s.prepareFile(path)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer runutil.CloseWithErrCapture(&err, fileCloser, "closing parquet file")
+
+	for rgN, rg := range rowGroups {
+		level.Debug(s.logger).Log("msg", "writing row group", "path", path, "row_group_number", rgN, "rows", rg.NumRows())
+
+		nInt64, err := s.writer.WriteRowGroup(rg)
+		if err != nil {
+			return 0, 0, err
+		}
+		n += uint64(nInt64)
+		numRowGroups += 1
+	}
+
+	if err := s.writer.Close(); err != nil {
+		return 0, 0, err
+	}
+
+	s.rowsFlushed += n
+
+	return n, numRowGroups, nil
+}
+
+func (s *profileStore) ingest(_ context.Context, profiles []*schemav1.Profile, lbs phlaremodel.Labels, profileName string, rewriter *rewriter) error {
+
+	var (
+		getters = make([]*profileGetter, len(profiles))
+	)
+
+	for pos := range profiles {
+		// rewrite elements
+		if err := s.helper.rewrite(rewriter, profiles[pos]); err != nil {
+			return err
+		}
+
+		// prepare getters
+		getters[pos] = &profileGetter{p: profiles[pos]}
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// do a binary search where to insert each profile
+	for pos, p := range profiles {
+		// check if row group is full
+		if s.cfg.MaxBufferRowCount > 0 && len(s.slice) >= s.cfg.MaxBufferRowCount ||
+			s.cfg.MaxRowGroupBytes > 0 && s.size.Load() >= s.cfg.MaxRowGroupBytes {
+			if err := s.cutRowGroup(); err != nil {
+				return err
+			}
+		}
+
+		// add profile to the index
+		s.index.addWithGetter(getters[pos], lbs, profileName)
+
+		// increase size of stored data
+		s.size.Add(s.helper.size(profiles[pos]))
+
+		// find correct position in the slice using binary search. The order needs to match the one from the schema
+		posProfile, _ := sort.Find(len(s.slice), func(i int) int {
+			// first compare the labels, if they don't match return
+			if cmp := phlaremodel.CompareLabelPairs(lbs, s.index.profilesPerFP[s.slice[i].SeriesFingerprint].lbs); cmp != 0 {
+				return cmp
+			}
+
+			// then compare timenanos, if they don't match return
+			if s.slice[i].TimeNanos < p.TimeNanos {
+				return 1
+			} else if s.slice[i].TimeNanos > p.TimeNanos {
+				return -1
+			}
+
+			// finally use ID as tie breaker
+			return bytes.Compare(p.ID[:], s.slice[i].ID[:])
+		})
+
+		// insert at the end of the slices
+		if len(s.slice) == posProfile {
+			s.slice = append(s.slice, p)
+			s.getters = append(s.getters, getters[pos])
+			continue
+		}
+
+		// make room at the correct position
+		s.slice = append(s.slice[:posProfile+1], s.slice[posProfile:]...)
```

Comment:
This can be quite heavy on the write path. I wonder why we need this in order when flushed files already provide this.

Reply:
We would need to go through a buffer to get the sorting as before. With only using the writer we would need to have it sorted before we pass it to the writer (when flushing a row group to disk).

In memory the sorting indeed doesn't provide any value. Alternatively we could just sort it once before flushing it to disk.

Let me compare those...

Reply:
Sorting once might be the way !